Feature/ex 3

nikiJava opened the pull request:
View Exercise 3.

nikiJava committed:
Implement news section.

nikiJava committed:
Add another layout (viewType) for criminal news.

nikiJava committed:
Add placeholder and error drawable into news list.

nikiJava committed:
Merge branch 'develop' into feature/ex-3

nikiJava committed:
Add CollapsingToolbar on news details screen.

nikiJava committed:
Reformat news list layout.

nikiJava committed:
Replace GridLayoutManager with StaggeredGridLayoutManager because of different items height.

nikiJava committed:
Remove duplicate keys in gradle properties.

nikiJava committed:
Remove redundant chars remained from merge conflict in styles.xml.

nikiJava committed:
Remove redundant strings.xml with the only Russian locale (ru-rRu) (result of branches merging).

nikiJava committed:
Add contentDescription for images in news items.

colriot commented on app/src/main/java/ru/nikijava/androidacademynewsapp/presentation/news/list/NewsListActivity.java:
```diff
@@ -0,0 +1,101 @@
+        RequestManager glide = Glide.with(this)
+                                    .setDefaultRequestOptions(options);
+        CompositeDelegateAdapter adapter = new CompositeDelegateAdapter.Builder<Item>()
+                .add(new NewsAdapter(this, glide, R.layout.item_news))
```
Спорное решение. Не вижу тут делегирования. Адаптеры одни и те же, только разметка меняется. При этом в `createViewHolder` всё равно идёт свич. Можно было бы сделать такой же свич в `onCreateViewHolder`, сделать общего предка у VH и в `onBindViewHolder` ничего не проверять. В общем, немного опять нагромождение абстракций ради абстракций.

colriot commented on app/src/main/java/ru/nikijava/androidacademynewsapp/presentation/news/list/NewsListActivity.java:
```diff
@@ -0,0 +1,101 @@
+        rvNewsList.setAdapter(adapter);
+
+        RecyclerView.LayoutManager layoutManager;
+        switch (getResources().getConfiguration().orientation) {
```
В задании был ещё пункт про разное отображение в зависимости от ширины экрана - не думал над этим?

colriot commented on app/src/main/res/layout/activity_news_details.xml:
```diff
@@ -0,0 +1,113 @@
+                android:layout_marginTop="@dimen/spacing_normal"
+                android:layout_marginEnd="@dimen/spacing_normal"
+                android:textAppearance="@style/TextAppearance.Details.Title"
+                app:layout_constraintEnd_toEndOf="parent"
```
Левые атрибуты от Констрейнтра остались

nikiJava replied:
Удалил лишние атрибуты.

colriot commented on app/src/main/res/layout/activity_news_list.xml:
```diff
@@ -0,0 +1,23 @@
+        android:layout_width="match_parent"
+        android:layout_height="match_parent"
+        android:clipToPadding="false"
+        android:contentDescription="@string/content_description_image_for_news"
```
Не нужно для RV такое поле. Более того, у тебя дескрипшн от IV

nikiJava replied:
Убрал ненужный дескрипшн.

colriot commented on app/src/main/res/layout/item_news.xml:
```diff
@@ -0,0 +1,86 @@
+            android:id="@+id/tvBody"
+            android:layout_width="0dp"
+            android:layout_height="wrap_content"
+            android:layout_marginTop="16dp"
```
Базовые димены хорошо бы вынести в `dimens.xml`

nikiJava replied:
Константы использовал только для паддингов. Для layout_margin константы не использовал, так как все вьюшки внутри ConstraintLayout располагаются.

colriot commented on app/src/main/res/layout/item_news.xml:
```diff
@@ -0,0 +1,86 @@
+        <ImageView
+            android:id="@+id/ivImage"
+            android:layout_width="100dp"
+            android:layout_height="0dp"
```
Сделал, чтобы оно растягивалось при ресайзе? Но я не вижу констрейнта на левый край вьюшки. Тогда почему так?

colriot replied:
Упустил то, что ширина вью фиксированная. Всё норм с точки зрения реализации, а с точки зрения UX - спорно - картинка будет вертикальная. Размер шрифта может менять, т.е и высотка текста, и высота картинки тоже, а ширина фикси. И ещё: когда будут короткие тексты, у тебя будет помногу пустого места внизу - возможно не есть гут, и данамическая высота была бы лучше.

nikiJava replied:
Сделал с помощью барьеров для картинки соотношение высоты к ширине как 1:1 - позволило избавиться от жёстко указанной ширины. На мой взгляд - это всё же минус, когда в одном и том же списке у одинаковых айтемов высота картинки будет разной, поскольку будет зависеть от размера заголовка и дескрипшена. Возможно, что это уже субъективной вопрос, но опыт пользования различными новостными приложениями да и просто приложениями, где есть список, подсказывает, что картинка всё должна быть одинаковой на всех айтемах. Да и желательно квадратной (1:1).

colriot commented on app/src/main/res/layout/item_news_big.xml:
```diff
@@ -0,0 +1,93 @@
+            android:id="@+id/tvTitle"
+            android:layout_width="0dp"
+            android:layout_height="wrap_content"
+            android:background="#b4c6c5c5"
```
Никогда не хардкодь цвета

nikiJava replied:
Убрал цвет константу.

colriot commented on app/src/main/res/layout/item_news_big.xml:
```diff
@@ -0,0 +1,93 @@
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_marginStart="16dp"
+            android:layout_marginTop="8dp"
```
Размеры то из констант, то хардкод

nikiJava replied:
Константы использовал только для паддингов. Для layout_margin константы не использовал, так как все вьюшки внутри ConstraintLayout располагаются.

colriot commented on delegateadapter/src/main/java/ru/nikijava/adapterdelegate/CompositeDelegateAdapter.java:
```diff
@@ -73,15 +74,16 @@ public void swapData(@NonNull final List<T> data) {
 
         public Builder() {
-            typeToAdapterMap = new SparseArray<>();
+            typeToAdapterMap = new HashMap<>();
```
Лучше юзать ArrayMap, для более эффективной работы по памяти. Т.к тебе не надо постоянно изменять мапу, а только несколько раз добавить в неё, а потом лишь читать.

nikiJava replied:
Заменил на ArrayMap.

colriot reviewed (changes requested):
Отревьюил

nikiJava committed:
Remove redundant kotlin gradle plugin.

nikiJava committed:
Extract androidx version of RecyclerView and CardView into separate variable.

nikiJava commented on app/src/main/java/ru/nikijava/androidacademynewsapp/presentation/news/details/NewsDetailsActivity.java:
```diff
@@ -0,0 +1,85 @@
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_news_details);
+        news = (News) getIntent().getSerializableExtra(NEWS_KEY);
+        setSupportActionBar(findViewById(R.id.toolbar));
```
Специально не проверял на null, поскольку таким образом можно замаскировать собственную или чужую ошибку (кто-то случайно поменял id у тулбара, или вовсе убрал без предупреждения), а так, по крэшу при тестировании я сразу ошибку увижу, что тулбар почему-то не найден на макете, увижу строчку, в которой ошибка.

colriot replied:
Подход имеет смысл. Зависит от перспективы и приоритетов.
Добро 👍  Классно, что оспариваешь комменты и аргументируешь!

nikiJava commented on app/src/main/java/ru/nikijava/androidacademynewsapp/DateFormatter.java:
```diff
@@ -0,0 +1,23 @@
+
+public class DateFormatter {
+
+    public CharSequence formatDateTime(Context context, Date dateTime) {
```
Если делать метод как static, то в таком случае и вызывать его нужно будет не через инстанс, а через класс. И как явную зависимость этот класс уже не объявишь потом - как такового смысла в этом не будет. Или такие классы с небольшой и просто логикой не стоит всегда указывать, как явную зависимость? Или суть в том, что у этого класса нет состояния и, следовательно, создавать инстанс смысла нет?

colriot replied:
Нет состояния, и он больше как Utils-класс. Если бы у тебя была возможность разные конфигурации сделать, подменять реализацию, то да - инстанс

nikiJava replied:
Исправил метод на static

nikiJava commented on app/src/main/java/ru/nikijava/androidacademynewsapp/presentation/news/details/NewsDetailsActivity.java:
```diff
@@ -0,0 +1,85 @@
+public class NewsDetailsActivity extends AppCompatActivity {
+
+    private static String NEWS_KEY = "news_key";
+    private final DateFormatter dateFormatter = new DateFormatter();
```
Согласен - нет смысла каждый раз создавать новый инстанс объекта при открытии экрана.

nikiJava committed:
Make DateFormatter field in news details screen static so don't create new instance of it every time.

nikiJava committed:
Extract background color from item_news_big into color resources.

nikiJava committed:
Replace HashMap with ArrayMap in CompositeDelegateAdapter for better memory efficiency

nikiJava commented on app/build.gradle:
```diff
@@ -17,18 +17,26 @@ android {
 
 dependencies {
+    implementation fileTree(dir: 'libs', include: ['*.jar'])
+    implementation "androidx.cardview:cardview:1.0.0"
```
Вынес в отдельную переменную.

nikiJava commented on app/build.gradle:
```diff
@@ -17,18 +17,26 @@ android {
 }
 
 dependencies {
+    implementation fileTree(dir: 'libs', include: ['*.jar'])
```
Убрал папку.

nikiJava committed:
Remove from build gradle compiling .jar files from /libs.

nikiJava committed:
Remove redundant contentDescription for RecyclerView in news list layout.

nikiJava committed:
Remove redundant attributes of ConstraintLayout in news details layout.

nikiJava commented on app/src/main/java/ru/nikijava/androidacademynewsapp/presentation/news/details/NewsDetailsActivity.java:
```diff
@@ -0,0 +1,85 @@
+        Glide.with(this).load(news.getImageUrl()).into(ivImage);
+    }
+
+    private void recomputeAvatarSize() {
```
Имеешь ввиду через веса?

colriot replied:
Да

nikiJava replied:
ImageView находится внутри CollapsingToolbar. В данной ситуации другого решения не вижу.

colriot commented on app/src/main/res/values/colors.xml:
```diff
@@ -4,4 +4,5 @@
     <color name="colorPrimaryDark">#ffd4a515</color>
     <color name="colorAccent">#ff962326</color>
     <color name="window_background">?android:attr/colorBackground</color>
+    <color name="background_transparent_grey">#b4c6c5c5</color>
```
Если ты прям акцент хочет на цвете сделать (и много где его юзать), то ок.
А вот если тут акцент скорее на назначении, типа "цвет фона карточки", то может лучше именно так его и назвать. Т.к он может поменяться, но не надо будет ещё и название менять (если он станет, например, синим). Но это не обязательно к исправлению, просто имей в виду.

colriot reviewed (approved):
В целом approve. Есть мелкие замечания в ответах на комменты, то не критичны

nikiJava committed:
Set DateFormatter's formatDateTime method as static

DateFormatter has no behaviour and so it is most preferably to not create redundant instances.

nikiJava committed:
Create separate color constant for big news item title background

nikiJava committed:
Implement ratio 1:1 for image in news item

nikiJava committed:
Set StaggeredGridLayoutManager for news list in all configurations

it helps to save scroll position in time of orientation changing.
